docs/20-design.md

```diff
@@ -0,0 +1,59 @@
+# Design: the Why of ContainerPilot
+
+## Why active service discovery?
+
+The goal of service discovery in cloud-native applications is to configure dependencies for our applications and update them as changes occur without manual intervention. There are two major approaches to solving this problem:
+
+- Insert a proxy between the application and the dependency. The proxy knows which dependency instances are available and implements load balancing among them.
+- Add dependency load balancing and dynamic configuration to the application itself.
+
+The first option is a common approach in the container ecosystem. It represents all the right reasons to outsource a problem: we get quick scalability with no need to modify the existing application. The proxy is a centralized place to manage access to dependencies, so all we have to do is point the client application at the proxy and go. The proxy itself can be as simple as HAProxy, or it can be a more complex proxy that can direct writes to a master and reads to the read replicas. This doesn't entirely solve the discovery problem but at least it centralizes the solution.
```

Comment:
> that can direct writes to a master and reads to the read replicas

Oops, should be "...that can direct writes to a primary and reads to the read replicas." Using "master" in that scenario is considered outdated language.